### PR TITLE
Semantics checks

### PR DESCRIPTION
**Pull request by rusya-ink**
- parameter
- neq operator
- neg operator
- add operator
- variable declaration
- integer literal
- greater operator

### CODE REVIEW

File: ast-nodes/variable-declaration.dart [hunk 1]
```diff
@@ -63,6 +63,8 @@ class VariableDeclaration extends Declaration {
   }
 
   void checkSemantics() {
-    // TODO: implement
+    this.type.checkSemantics();
+    this.value.checkSemantics();
+    ensureType(this.value, this.type);
```
Comment (GneyHabub) [inline ast-nodes/variable-declaration.dart]: `ensureType` produces no meaningful side effect, it returns a value. Also, `this.value` and `this.type` might both be `null` and `ensureType` will crash if you pass null as an argument.

File: ast-nodes/expressions/operators/greater-operator.dart [hunk 2]
```diff
@@ -30,6 +30,23 @@ class GreaterOperator extends BinaryRelation implements Comparison {
...
+      this.leftOperand = ensureType(this.leftOperand, RealType());
+      this.rightOperand = ensureType(this.rightOperand, RealType());
+    } else if (leftType is IntegerType || rightType is IntegerType) {
+      this.leftOperand = ensureType(this.leftOperand, IntegerType());
+      this.rightOperand = ensureType(this.rightOperand, IntegerType());
+    } else {
+      if (leftType != rightType) {
+        throw SemanticError(this, "Cannot compare objects of different types");
+      }
+    }
```
Comment (GneyHabub) [inline ast-nodes/expressions/operators/greater-operator.dart]: You have the same mistake as I had. We can compare bools, arrays and records only for equality, we cannot say if one is greater then another.

File: ast-nodes/expressions/operators/neg-operator.dart [hunk 3]
```diff
@@ -25,6 +25,7 @@ class NegOperator extends UnaryRelation implements Primary {
   }
 
   void checkSemantics() {
-    // TODO: implement
+    this.operand.checkSemantics();
+    ensureType(this.operand, this.resultType);
```
Comment (GneyHabub) [inline ast-nodes/expressions/operators/neg-operator.dart]: You are the one that has to set the resulting type, you cannot use it. And the same mistake with `ensureType` as in one of the previous commits.

**Review by GneyHabub — CHANGES_REQUESTED** [review]

File: ast-nodes/expressions/operators/neg-operator.dart [hunk 4]
```diff
@@ -26,6 +26,13 @@ class NegOperator extends UnaryRelation implements Primary {
 
   void checkSemantics() {
     this.operand.checkSemantics();
-    ensureType(this.operand, this.resultType);
+    if (operand is IntegerType) {
```
Comment (illright) [inline ast-nodes/expressions/operators/neg-operator.dart]: You need to check `operand.resultType`, not just `operand`

File: ast-nodes/variable-declaration.dart [hunk 5]
```diff
@@ -65,6 +65,21 @@ class VariableDeclaration extends Declaration {
   void checkSemantics() {
     this.type.checkSemantics();
     this.value.checkSemantics();
```
Comment (illright) [inline ast-nodes/variable-declaration.dart]: Use the optional chaining here (`?.`), both value and the type may be `null`.

File: ast-nodes/variable-declaration.dart [hunk 6]
```diff
@@ -65,6 +65,21 @@ class VariableDeclaration extends Declaration {
...
-    ensureType(this.value, this.type);
+    if (this.type != null && this.value != null) {
```
Comment (illright) [inline ast-nodes/variable-declaration.dart]: Okay, I really don't understand what is going on in this piece of code. Let me tell you how I see this semantic check:
* if type is not null, check its semantics
* if value is not null, check its semantics
* if type is null, assign `this.type = this.value.returnType`
* if type is not null, ensure the value is convertible to the type

I don't know what are all those huge if statements about

**Review by illright — CHANGES_REQUESTED** [review]

**Review by GneyHabub — APPROVED** [review]

**Review by illright — APPROVED** [review]

**Pull request merged**